mmdet3d/models/dense_heads/groupfree3d_head.py

```diff
@@ -0,0 +1,1000 @@
+
+            # calculate objectness loss
+            obj_score = bbox_preds[f'obj_scores{suffix}'].transpose(2,
+                                                                    1).reshape(
+                                                                        -1, 1)
+            objectness_loss = self.objectness_loss(
+                obj_score,
+                1 - objectness_targets.reshape(-1),
+                objectness_weights.reshape(-1),
+                avg_factor=batch_size)
+
+            objectness_loss_sum += objectness_loss
+
+            # calculate center loss
+            box_loss_weights_expand = box_loss_weights.unsqueeze(-1).repeat(
+                1, 1, 3)
+            center_loss = self.center_loss(
+                bbox_preds[f'center{suffix}'],
+                assigned_center_targets,
+                weight=box_loss_weights_expand)
+
+            center_loss_sum += center_loss
+
+            # calculate direction class loss
+            dir_class_loss = self.dir_class_loss(
+                bbox_preds[f'dir_class{suffix}'].transpose(2, 1),
+                dir_class_targets,
+                weight=box_loss_weights)
+
+            dir_class_loss_sum += dir_class_loss
+
+            # calculate direction residual loss
+            heading_label_one_hot = size_class_targets.new_zeros(
+                (batch_size, proposal_num, self.num_dir_bins))
+            heading_label_one_hot.scatter_(2, dir_class_targets.unsqueeze(-1),
+                                           1)
+            dir_res_norm = torch.sum(
+                bbox_preds[f'dir_res_norm{suffix}'] * heading_label_one_hot,
+                -1)
+            dir_res_loss = self.dir_res_loss(
+                dir_res_norm, dir_res_targets, weight=box_loss_weights)
+
+            dir_res_loss_sum += dir_res_loss
+
+            if self.size_cls_agnostic:
+                # calculate class-agnostic size loss
+                size_reg_loss = self.size_reg_loss(
+                    bbox_preds[f'size{suffix}'],
+                    assigned_size_targets,
+                    weight=box_loss_weights_expand)
+
+                size_reg_loss_sum += size_reg_loss
+
+            else:
+                # calculate size class loss
+                size_class_loss = self.size_class_loss(
+                    bbox_preds[f'size_class{suffix}'].transpose(2, 1),
+                    size_class_targets,
+                    weight=box_loss_weights)
+
+                size_class_loss_sum += size_class_loss
+
+                # calculate size residual loss
+                one_hot_size_targets = size_class_targets.new_zeros(
+                    (batch_size, proposal_num, self.num_sizes))
+                one_hot_size_targets.scatter_(2,
+                                              size_class_targets.unsqueeze(-1),
+                                              1)
+                one_hot_size_targets_expand = one_hot_size_targets.unsqueeze(
```

Comment:
Expand is better, since repeat copies but expand doesn't when the before-expansion dim size is 1